Core/Src/vcu.c

```diff
@@ -52,6 +68,153 @@ void can_buffer_handling_loop()
 	service_can_tx_hardware(hcan);
 }
 
+void control_cooling() {
+	// TODO be smart about when turning on pump and fan
+	if(brakeTempFrontLeft_C.data == 1 || brakeTempFrontRight_C.data == 1){
+		HAL_GPIO_WritePin(PUMP_GPIO_Port, PUMP_Pin, GPIO_PIN_SET);
+		HAL_GPIO_WritePin(FAN_GPIO_Port, FAN_Pin, GPIO_PIN_SET);
+	}
+
+
+	// TODO: Dynamic cooling system management
+}
+
+void handle_CAN() {
+
+}
+
+void handle_inputs() {
+
+}
+
+void brake_light() {
+	// If brakePressureRear_psi.data is over 0, the brake is pressed
+	if(brakePressureRear_psi.data > 0) {
+		// If the brake light isn't on already, we need to turn it on
+		if (HAL_GPIO_ReadPin(BRK_LT_GPIO_Port, BRK_LT_Pin) == GPIO_PIN_RESET) {
+			HAL_GPIO_WritePin(BRK_LT_GPIO_Port, BRK_LT_Pin, GPIO_PIN_SET);
+		}
+	}
+	// Else, the brake isn't pressed
+	else {
+		// If the brake light isn't pressed but the light is still on, we need to turn it off
+		if (HAL_GPIO_ReadPin(BRK_LT_GPIO_Port, BRK_LT_Pin) == GPIO_PIN_SET) {
+			HAL_GPIO_WritePin(BRK_LT_GPIO_Port, BRK_LT_Pin, GPIO_PIN_RESET);
+		}
+	}
+	return;
+}
+
+void handle_inv(DRIVE_STATE_t *curr_state) {
+	U8 msg_data[8];
+	U16 torque_req = 0;
+
+	//update_inv_state();
+
+	// find the state of the inverter based on the status of the ECU
+//	if (inv_states.inv_enable_lockout)
+//	{
+//		inv_ctrl_state = INV_LOCKOUT;
+//	}
+//	else if (!inv_states.inv_en_state)
+//	{
+//		// not locked out, not enabled
+//		inv_ctrl_state = INV_DISABLED;
+//
+//		// disable the state
+//		*curr_state = NOT_READY_STATE;
+//	}
+//	else
+//	{
+//		inv_ctrl_state = INV_ENABLED;
+//	}
+
+	// if the inverter is locked out, we must first disable and enable the motor
+	switch (*curr_state) {
+		case INV_LOCKOUT:
+			// send the disable then enable command with some delay
+			CAN_cmd.inverter_en = 0;
+			build_cmd_msg(msg_data, &CAN_cmd);
+			send_can_message(COMMAND_MSG_ID, 8, msg_data);
+
+			// also clear all faults when first starting up
+			rw_cmd.param_addr = FAULT_CLEAR;
+			rw_cmd.read_or_write = WRITE_CMD;
+			rw_cmd.data = 0;
+			build_param_cmd_msg(msg_data, &rw_cmd);
+			send_can_message(PARAM_RW_CMD_ID, 8, msg_data);
+			break;
+
+		case INV_DISABLED:
+			// try to enable the motor
+			if (*curr_state == INV_READY)
+			{
+				CAN_cmd.inverter_en = 1;
+				build_cmd_msg(msg_data, &CAN_cmd);
+				send_can_message(COMMAND_MSG_ID, 8, msg_data);
+			}
+			break;
+
+		case INV_ENABLED:
+			// actually run the motor
+
+			// calculate the torque we want from the motor. Right now we are linear
+			if (brakePressureRear_psi.data <= BRAKE_PRESSURE_BREAK_THRESH_psi)
+			{
+				// Old command was "float curr_pp = vcu_apps2.data;"
+				float curr_pp =pedalPosition2_mm.data;
+				if (curr_pp >= APPS2_MAX_TRAVEL_mm)
+				{
+					torque_req = MAX_TORQUE_Nm;
```

Comment:
How much of this have you been able to validate so far?

Reply:
Almost none, just made sure that there were no errors.